src/hooks/useAddressHistory.ts

```diff
@@ -0,0 +1,77 @@
+import { atom, useRecoilState } from "recoil";
+import { useAddressInput } from "hooks/useAddressInput";
+import { useNavigate } from "react-router-dom";
+import { isValidAddress } from "utils";
+import useNotification from "hooks/useNotification";
+import { ROUTES } from "consts";
+import { recoilPersist } from "recoil-persist";
+import { Address } from "ton";
+import { useEffect } from "react";
+import { useJettonAddress } from "hooks/useJettonAddress";
+
+const { persistAtom } = recoilPersist();
+
+const addressHistoryState = atom({
+  key: "addressHistory",
+  default: [] as string[],
+  effects_UNSTABLE: [persistAtom],
+});
+
+export function useAddressHistory() {
+  const [addresses, setAddresses] = useRecoilState(addressHistoryState);
+  const { setActive, setValue, addressInput } = useAddressInput();
+  const navigate = useNavigate();
+  const { showNotification } = useNotification();
+  const { id } = useJettonAddress();
+
+  const resetAddresses = () => {
+    setAddresses([]);
+    setActive(false);
+  };
+
+  const removeAddress = (address: string) =>
+    setAddresses((prev: string[]) => [...prev.filter((a) => a !== address)]);
+
+  const onAddressClick = (address: string) => {
+    setActive(false);
+    setValue("");
+
+    setAddresses((prev: string[]) => [address, ...prev.filter((a) => a !== address)]);
```

Comment:
there are 3 different code paths that add an address, I'd extract it to a function so we don't miss out on functionality (for example here slicing 20 addresses is missing I think).